cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/harbor_api_v2_repo_lister.go

```diff
@@ -0,0 +1,298 @@
+		return nil, err
+	}
+}
+
+// buildHarborRegistryBaseURL builds the URL for accessing the base API.
+func buildHarborRegistryBaseURL(plainHTTP bool, ref orasregistryv2.Reference) string {
+	scheme := "https"
+	if plainHTTP {
+		scheme = "http"
+	}
+	return fmt.Sprintf("%s://%s/api/v2.0", scheme, ref.Host())
+}
+
+func harborProjectNameFromURL(url url.URL) (string, error) {
+	path := strings.TrimPrefix(url.Path, "/")
+	segments := strings.SplitN(path, "/", 2)
+	if len(segments) > 0 {
+		return segments[0], nil
+	} else {
+		return "", fmt.Errorf("unexpected URL format: [%s]", url.String())
+	}
+}
+
+func harborDoHttpRequest(ctx context.Context, method string, cred orasregistryauthv2.Credential, url string) (*http.Response, error) {
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
+		return nil, err
+	}
+	if cred.Username != "" && cred.Password != "" {
+		req.SetBasicAuth(cred.Username, cred.Password)
+	}
+	return httpclient.New().Do(req)
+}
+
+// maxErrorBytes specifies the default limit on how many response bytes are
+// allowed in the server's error response.
+// A typical error message is around 200 bytes. Hence, 8 KiB should be
+// sufficient.
+var maxErrorBytes int64 = 8 * 1024 // 8 KiB
+
+// requestError contains a single error.
+type requestError struct {
+	Code    string `json:"code"`
+	Message string `json:"message"`
+}
+
+// requestErrors is a bundle of requestError.
+type requestErrors []requestError
+
+// Error returns a error string describing the error.
+func (e requestError) Error() string {
+	code := strings.Map(func(r rune) rune {
+		if r == '_' {
+			return ' '
+		}
+		return unicode.ToLower(r)
+	}, e.Code)
+	if e.Message == "" {
+		return code
+	}
+	return fmt.Sprintf("%s: %s", code, e.Message)
+}
+
+// Error returns a error string describing the error.
+func (errs requestErrors) Error() string {
+	switch len(errs) {
+	case 0:
+		return "<nil>"
+	case 1:
+		return errs[0].Error()
+	}
+	var errmsgs []string
+	for _, err := range errs {
+		errmsgs = append(errmsgs, err.Error())
+	}
+	return strings.Join(errmsgs, "; ")
+}
+
+type harborRepoModel struct {
+	Name string `json:"name"`
+}
+
+func parseHarborProjectRepositoriesResponse(resp *http.Response) ([]string, error) {
+	contentType := resp.Header.Get("Content-Type")
+	if !strings.HasPrefix(contentType, "application/json") {
+		return nil, fmt.Errorf("unexpected response content type: [%s]", contentType)
+	}
+	var respRepos []harborRepoModel
+	if err := json.NewDecoder(resp.Body).Decode(&respRepos); err != nil {
+		return nil, err
+	} else {
+		repos := []string{}
+		for _, r := range respRepos {
+			repos = append(repos, r.Name)
+		}
+		return repos, nil
+	}
+}
+
+// parseHarborErrorResponse parses the error returned by the remote registry.
+func parseHarborErrorResponse(resp *http.Response) error {
+	var errmsg string
+	var body struct {
+		Errors requestErrors `json:"errors"`
+	}
+	lr := io.LimitReader(resp.Body, maxErrorBytes)
```

Comment:
Just curious as it seems a good idea for preventing some buffer overflow attacks (and we aren't doing it in the rest of the codebase, AFAIK). 
Do you think we should limit the read operations in other places?

Reply:
I think in general it is a good idea, provided you can make this assumption. Normally, you can't really assume anything about the size of the payload you're dealing with. So whatever limit/boundary X you pick, I will probably find a valid use case where receiving X+1 bytes is valid